plugins/thunderbolt/fu-plugin-thunderbolt.c

```diff
@@ -67,3 +67,32 @@ fu_plugin_startup(FuPlugin *plugin, GError **error)
 {
 	return fu_plugin_thunderbolt_safe_kernel(plugin, error);
 }
+
+gboolean
+fu_plugin_composite_prepare (FuPlugin *plugin,
+			     GPtrArray *devices,
+			     GError **error)
+{
+	for (guint i = 0; i < devices->len; i++) {
+		FuDevice *device = g_ptr_array_index (devices, i);
+		if (fu_device_has_flag(device, FWUPD_DEVICE_FLAG_NO_AUTO_REMOVE)) {
+			return fu_thunderbolt_device_open(device,error);
+		}
+	}
+	return TRUE;
+}
+
+
+gboolean
+fu_plugin_composite_cleanup	(FuPlugin	*plugin,
+							 GPtrArray	*devices,
+							 GError		**error)
```

Comment:
the spacing is funky here -- are you not using the precommit and clang-format rules?

Reply:
Thanks @hughsie  am using eclipse , I was developing on chrome till now but will need to format code compile code on native ubuntu fix all warnings etc , will post the updates soon here

Reply:
Run contrib/setup

Reply:
./contrib/ci/generate_dependencies.py -o ubuntu | sudo xargs apt install -y
./contrib/setup
meson build -Ddocs=none
after above steps , did you face below any work around please ? 
plugins/tpm/meson.build:46:2: ERROR: Bad source of type 'SharedModule' in target 'tpm-self-test'.

Reply:
I think you need to pull/rebase on latest main.  That's fixed in https://github.com/fwupd/fwupd/commit/3e2e0aa61600f73433b149a183befddf5ab1bcef

Reply:
Thats fixed, "meson build -Ddocs=none" configures the build right , how do we compile code in this env ?

Reply:
This will build and run test suite:

```
ninja -C build test
```

Reply:
from https://github.com/fwupd/fwupd/wiki/Compilation using "ninja -C build"

Reply:
I have fixed a failing test case since usb4_port1 dir was not available in https://github.com/fwupd/fwupd/pull/4044/commits/2442c83a5f2b5e59bb9069c4ef46571a5cf7960e

Reply:
new PR is here https://github.com/fwupd/fwupd/pull/4044

Reply:
not sure how to update this branch as it needs fast forward now, hence created a new https://github.com/fwupd/fwupd/pull/4044